In [1]:
import numpy as np
import pandas as pd

In [2]:
sales_data = pd.read_csv("sales_data.csv")
customer_data = pd.read_csv("customer_data.csv")

**Q1. Load both sales and customer dataset into jupyter.**

- Display first 10 rows of sales_data.
- Display last 15 rows of customer_data.

In [5]:
sales_data.head(10)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM
4,I337046,C189076,Books,4,60.60,24-10-2021,Kanyon
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul
6,I121056,C151197,Cosmetics,1,40.66,13-03-2022,Istinye Park
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul
8,I293455,C159642,Clothing,3,900.24,04-11-2021,Metrocity
9,I326945,C283361,Clothing,2,600.16,22-08-2021,Kanyon


In [6]:
customer_data.tail(15)

,customer_id,gender,age,payment_method
99442,C131810,Female,45.0,Debit Card
99443,C223470,Male,41.0,Debit Card
99444,C119287,Male,24.0,Cash
99445,C270290,Female,35.0,Cash
99446,C107209,Female,32.0,Cash
99447,C288090,Female,37.0,Cash
99448,C231387,Female,65.0,Credit Card
99449,C953724,Male,65.0,Cash
99450,C226974,Female,28.0,Cash
99451,C513603,Male,50.0,Cash


**Q2. Display the shape and size of both the datasets.**
- Display their column names.

In [10]:
customer_data.columns.to_list()

['customer_id', 'gender', 'age', 'payment_method']

In [11]:
sales_data.columns.to_list()

['invoice_no',
 'customer_id',
 'category',
 'quantity',
 'price',
 'invoice_date',
 'shopping_mall']

**Q3. Identify the summary of sales dataset and display each columns datatype and suggest any
conversion required?**

In [12]:
sales_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 99457 entries, 0 to 99456
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   invoice_no     99457 non-null  str    
 1   customer_id    99457 non-null  str    
 2   category       99457 non-null  str    
 3   quantity       99457 non-null  int64  
 4   price          99457 non-null  float64
 5   invoice_date   99457 non-null  str    
 6   shopping_mall  99457 non-null  str    
dtypes: float64(1), int64(1), str(5)
memory usage: 5.3 MB


In [13]:
sales_data.dtypes

invoice_no           str
customer_id          str
category             str
quantity           int64
price            float64
invoice_date         str
shopping_mall        str
dtype: object

As checked all the colums/features are showing correct data types so there is no conversion required.

**Q4. Identify the summary of customers dataset and display each columns datatype and suggest
any conversion required?**

In [15]:
customer_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 99457 entries, 0 to 99456
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   customer_id     99457 non-null  str    
 1   gender          99457 non-null  str    
 2   age             99338 non-null  float64
 3   payment_method  99457 non-null  str    
dtypes: float64(1), str(3)
memory usage: 3.0 MB


In [16]:
customer_data.dtypes

customer_id           str
gender                str
age               float64
payment_method        str
dtype: object

As checked all the colums are having correct data type as per there values so no conversion required 

**Q5. Identify the unique_values and value counts of category in sales_data and payment_method
of customers_data ?**

In [20]:
sales_data['category'].unique()

<StringArray>
[       'Clothing',           'Shoes',           'Books',       'Cosmetics',
 'Food & Beverage',            'Toys',      'Technology',        'Souvenir']
Length: 8, dtype: str

In [21]:
sales_data['category'].value_counts()

category
Clothing           34487
Cosmetics          15097
Food & Beverage    14776
Toys               10087
Shoes              10034
Souvenir            4999
Technology          4996
Books               4981
Name: count, dtype: int64

In [22]:
customer_data['payment_method'].unique()

<StringArray>
['Credit Card', 'Debit Card', 'Cash']
Length: 3, dtype: str

In [23]:
customer_data['payment_method'].value_counts()

payment_method
Cash           44447
Credit Card    34931
Debit Card     20079
Name: count, dtype: int64

**Q6. check for the missing values in sales_data and customers data?**

In [27]:
sales_data.isnull().sum()

invoice_no       0
customer_id      0
category         0
quantity         0
price            0
invoice_date     0
shopping_mall    0
dtype: int64

there is no missing value in the sales_data dataframe

In [29]:
customer_data.isnull().sum()

customer_id         0
gender              0
age               119
payment_method      0
dtype: int64

there are 119 missing values in age column. there index is below

In [34]:
l = []
index = 0
for i in customer_data['age'].isnull():
    if i == True:
        l.append(index)
    index += 1
print("index of empty values")
print(l)
len(l)

    

index of empty values
[91, 92, 93, 94, 95, 97199, 97200, 97201, 97202, 97203, 97204, 97205, 97206, 97207, 97208, 97209, 97210, 97211, 97212, 97213, 97214, 97215, 97216, 97217, 97218, 97219, 97220, 97221, 97222, 97223, 97224, 97225, 97226, 97227, 97228, 97229, 97230, 97231, 97232, 97233, 97234, 97235, 97236, 97237, 97238, 97239, 97240, 97241, 97242, 97243, 97244, 97245, 97246, 97247, 97248, 97249, 97250, 97251, 97252, 97253, 97254, 97255, 97256, 97257, 97258, 97259, 97260, 97261, 97262, 97263, 97264, 97265, 97266, 97267, 97268, 97269, 97270, 97271, 97272, 97273, 97274, 97275, 97276, 97277, 97278, 97279, 97280, 97281, 97282, 97283, 97284, 97285, 97286, 97287, 97288, 97289, 97290, 97291, 97292, 97293, 97294, 97295, 97296, 97297, 97298, 97299, 97300, 97301, 97302, 97303, 97304, 97305, 97306, 97307, 97308, 97309, 97310, 97311, 97312]


119

In [35]:
customer_data.iloc[91]

customer_id          C157070
gender                Female
age                      NaN
payment_method    Debit Card
Name: 91, dtype: object

**Q7. Merge the sales and customers data_sets and set customer_id as index?**

In [36]:
all_data = pd.merge(sales_data,customer_data,on='customer_id')
all_data

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul,Male,21.0,Debit Card
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card
4,I337046,C189076,Books,4,60.60,24-10-2021,Kanyon,Female,53.0,Cash
...,...,...,...,...,...,...,...,...,...,...
99452,I219422,C441542,Souvenir,5,58.65,21-09-2022,Kanyon,Female,45.0,Credit Card
99453,I325143,C569580,Food & Beverage,2,10.46,22-09-2021,Forum Istanbul,Male,27.0,Cash
99454,I824010,C103292,Food & Beverage,2,10.46,28-03-2021,Metrocity,Male,63.0,Debit Card
99455,I702964,C800631,Technology,4,4200.00,16-03-2021,Istinye Park,Male,56.0,Cash


In [66]:
all_data.set_index("customer_id")

,invoice_no,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
customer_id,,,,,,,,,
C241288,I138884,Clothing,5.0,1500.40,05-08-2022,Kanyon,NaN,NaN,NaN
C111565,I317333,Shoes,3.0,1800.51,12-12-2021,Forum Istanbul,NaN,NaN,NaN
C266599,I127801,Clothing,1.0,300.08,09-11-2021,Metrocity,NaN,NaN,NaN
C988172,I173702,Shoes,5.0,3000.85,16-05-2021,Metropol AVM,NaN,NaN,NaN
C189076,I337046,Books,4.0,60.60,24-10-2021,Kanyon,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
C441542,NaN,NaN,NaN,NaN,NaN,NaN,Female,45.0,Credit Card
C569580,NaN,NaN,NaN,NaN,NaN,NaN,Male,27.0,Cash
C103292,NaN,NaN,NaN,NaN,NaN,NaN,Male,63.0,Debit Card


**Q8. Extract the category and quantity columns along with all rows**
- using the loc functio
- using iloc function extract only the last 50 rows.

In [37]:
all_data[['category','quantity']]

,category,quantity
0,Clothing,5
1,Shoes,3
2,Clothing,1
3,Shoes,5
4,Books,4
...,...,...
99452,Souvenir,5
99453,Food & Beverage,2
99454,Food & Beverage,2
99455,Technology,4


In [38]:
all_data.iloc[-50::]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
99407,I458636,C156575,Clothing,3,900.24,20-08-2021,Istinye Park,Female,67.0,Cash
99408,I949512,C217355,Clothing,2,600.16,21-02-2021,Metrocity,Female,47.0,Credit Card
99409,I591543,C283521,Toys,4,143.36,25-07-2021,Metropol AVM,Male,28.0,Credit Card
99410,I171613,C333571,Toys,2,71.68,18-10-2021,Mall of Istanbul,Female,22.0,Credit Card
99411,I231116,C210000,Cosmetics,4,162.64,12-08-2021,Zorlu Center,Female,33.0,Debit Card
99412,I637813,C241121,Toys,4,143.36,03-10-2021,Zorlu Center,Female,67.0,Cash
99413,I120044,C327276,Technology,1,1050.00,26-05-2021,Mall of Istanbul,Female,24.0,Debit Card
99414,I166977,C206152,Cosmetics,5,203.30,05-02-2022,Kanyon,Female,34.0,Credit Card
99415,I225127,C221559,Clothing,2,600.16,04-07-2021,Metropol AVM,Male,33.0,Credit Card
99416,I143010,C922388,Food & Beverage,5,26.15,22-12-2021,Cevahir AVM,Male,55.0,Cash


In [39]:
l=[]
for i in range (99407,99457):
    l.append(i)
print(l)
all_data.loc[l]

[99407, 99408, 99409, 99410, 99411, 99412, 99413, 99414, 99415, 99416, 99417, 99418, 99419, 99420, 99421, 99422, 99423, 99424, 99425, 99426, 99427, 99428, 99429, 99430, 99431, 99432, 99433, 99434, 99435, 99436, 99437, 99438, 99439, 99440, 99441, 99442, 99443, 99444, 99445, 99446, 99447, 99448, 99449, 99450, 99451, 99452, 99453, 99454, 99455, 99456]


,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
99407,I458636,C156575,Clothing,3,900.24,20-08-2021,Istinye Park,Female,67.0,Cash
99408,I949512,C217355,Clothing,2,600.16,21-02-2021,Metrocity,Female,47.0,Credit Card
99409,I591543,C283521,Toys,4,143.36,25-07-2021,Metropol AVM,Male,28.0,Credit Card
99410,I171613,C333571,Toys,2,71.68,18-10-2021,Mall of Istanbul,Female,22.0,Credit Card
99411,I231116,C210000,Cosmetics,4,162.64,12-08-2021,Zorlu Center,Female,33.0,Debit Card
99412,I637813,C241121,Toys,4,143.36,03-10-2021,Zorlu Center,Female,67.0,Cash
99413,I120044,C327276,Technology,1,1050.00,26-05-2021,Mall of Istanbul,Female,24.0,Debit Card
99414,I166977,C206152,Cosmetics,5,203.30,05-02-2022,Kanyon,Female,34.0,Credit Card
99415,I225127,C221559,Clothing,2,600.16,04-07-2021,Metropol AVM,Male,33.0,Credit Card
99416,I143010,C922388,Food & Beverage,5,26.15,22-12-2021,Cevahir AVM,Male,55.0,Cash


**Q9. Use .iloc to select the second row of the sales_data.**
- Retrieve the 5th to 10th rows (inclusive) and the 3rd column (index 2) in the
customer_data.

In [40]:
sales_data.iloc[1]

invoice_no              I317333
customer_id             C111565
category                  Shoes
quantity                      3
price                   1800.51
invoice_date         12-12-2021
shopping_mall    Forum Istanbul
Name: 1, dtype: object

In [41]:
sales_data.iloc[5:11]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul
6,I121056,C151197,Cosmetics,1,40.66,13-03-2022,Istinye Park
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul
8,I293455,C159642,Clothing,3,900.24,04-11-2021,Metrocity
9,I326945,C283361,Clothing,2,600.16,22-08-2021,Kanyon
10,I306368,C240286,Food & Beverage,2,10.46,25-12-2022,Metrocity


In [42]:
customer_data.iloc[:,2]

0        28.0
1        21.0
2        20.0
3        66.0
4        53.0
         ... 
99452    45.0
99453    27.0
99454    63.0
99455    56.0
99456    36.0
Name: age, Length: 99457, dtype: float64

**Q10. Select all rows where: category is not "Books".**

In [43]:
all_data[all_data['category'] != 'Books']

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul,Male,21.0,Debit Card
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card
...,...,...,...,...,...,...,...,...,...,...
99452,I219422,C441542,Souvenir,5,58.65,21-09-2022,Kanyon,Female,45.0,Credit Card
99453,I325143,C569580,Food & Beverage,2,10.46,22-09-2021,Forum Istanbul,Male,27.0,Cash
99454,I824010,C103292,Food & Beverage,2,10.46,28-03-2021,Metrocity,Male,63.0,Debit Card
99455,I702964,C800631,Technology,4,4200.00,16-03-2021,Istinye Park,Male,56.0,Cash


**Q11. Find all transactions where:The category is "Clothing" and the quantity is greater than 3.**

In [44]:
all_data[(all_data['category'] == 'Clothing') & (all_data['quantity'] > 3)]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card
33,I109053,C232624,Clothing,4,1200.32,11-07-2021,Metrocity,Male,51.0,Debit Card
36,I147062,C245456,Clothing,5,1500.40,21-06-2022,Kanyon,Male,43.0,Credit Card
47,I205366,C241642,Clothing,4,1200.32,07-11-2022,Zorlu Center,Female,43.0,Debit Card
...,...,...,...,...,...,...,...,...,...,...
99393,I145718,C331158,Clothing,5,1500.40,03-09-2022,Mall of Istanbul,Female,64.0,Cash
99404,I311155,C651683,Clothing,5,1500.40,29-01-2021,Kanyon,Male,18.0,Debit Card
99430,I336082,C316368,Clothing,5,1500.40,26-04-2021,Mall of Istanbul,Male,26.0,Cash
99436,I579905,C265057,Clothing,5,1500.40,27-03-2022,Cevahir AVM,Female,57.0,Cash


**Q12. Display the customers who made purchases at the "Mall of Istanbul" with the following
columns:
customer_id, age, category, quantity, and price.**

In [45]:
all_data[all_data['shopping_mall'] == 'Mall of Istanbul'][['customer_id','age', 'category', 'quantity', 'price']]

,customer_id,age,category,quantity,price
7,C176086,32.0,Clothing,2,600.16
26,C289625,25.0,Cosmetics,5,203.30
30,C159164,66.0,Toys,3,107.52
46,C317818,19.0,Cosmetics,3,121.98
60,C336576,64.0,Clothing,5,1500.40
...,...,...,...,...,...
99431,C131280,26.0,Cosmetics,4,162.64
99432,C289915,41.0,Books,2,30.30
99443,C223470,41.0,Cosmetics,4,162.64
99444,C119287,24.0,Toys,1,35.84


**Q13. Sort the sales data by price in descending order.
Filter customers aged between 25 and 40.**

In [46]:
sales_data.sort_values(by='price',ascending=False)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall
65681,I127941,C281698,Technology,5,5250.00,14-10-2022,Kanyon
17073,I428872,C248633,Technology,5,5250.00,10-04-2021,Istinye Park
31520,I213835,C509633,Technology,5,5250.00,19-03-2021,Mall of Istanbul
31509,I104876,C151588,Technology,5,5250.00,25-05-2021,Kanyon
89707,I392959,C324172,Technology,5,5250.00,22-12-2022,Istinye Park
...,...,...,...,...,...,...,...
77093,I326346,C422842,Food & Beverage,1,5.23,30-04-2021,Forum Istanbul
77082,I103004,C560575,Food & Beverage,1,5.23,04-02-2022,Metropol AVM
67569,I107064,C864821,Food & Beverage,1,5.23,13-02-2022,Metropol AVM
67844,I334331,C756260,Food & Beverage,1,5.23,17-06-2022,Kanyon


In [47]:
customer_data[customer_data['age'].between(25,40)]

,customer_id,gender,age,payment_method
0,C241288,Female,28.0,Credit Card
5,C657758,Female,28.0,Credit Card
7,C176086,Female,32.0,Credit Card
10,C240286,Female,36.0,Cash
11,C191708,Female,29.0,Credit Card
...,...,...,...,...
99446,C107209,Female,32.0,Cash
99447,C288090,Female,37.0,Cash
99450,C226974,Female,28.0,Cash
99453,C569580,Male,27.0,Cash


**Q14. Create a new column named total_spending in the merged dataset. This column should
calculate the total spending per transaction by multiplying the quantity and price columns.
Display the first 10 rows of the updated DataFrame to verify the result
List customers who shopped in "Mall of Istanbul" with purchases over $500.**

In [48]:
all_data['total_spending']= 0.0

In [49]:
all_data.tail(1)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending
99456,I232867,C273973,Souvenir,3,35.19,15-10-2022,Mall of Istanbul,Female,36.0,Credit Card,0.0


In [50]:
for i in range(0,99456):
    all_data.loc[i,'total_spending'] = all_data.loc[i,'quantity'] * all_data.loc[i,'price']

In [51]:
all_data[(all_data['shopping_mall']=="Mall of Istanbul") & (all_data['total_spending'] > 500)].head(10)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul,Female,32.0,Credit Card,1200.32
26,I196704,C289625,Cosmetics,5,203.30,20-04-2022,Mall of Istanbul,Female,25.0,Credit Card,1016.50
60,I133061,C336576,Clothing,5,1500.40,09-06-2022,Mall of Istanbul,Male,64.0,Credit Card,7502.00
63,I157285,C123427,Clothing,4,1200.32,06-03-2022,Mall of Istanbul,Male,66.0,Cash,4801.28
69,I883721,C236859,Technology,5,5250.00,19-11-2021,Mall of Istanbul,Female,44.0,Credit Card,26250.00
87,I253242,C309438,Shoes,5,3000.85,07-04-2022,Mall of Istanbul,Male,26.0,Cash,15004.25
106,I224371,C204174,Clothing,5,1500.40,01-02-2022,Mall of Istanbul,Male,34.0,Credit Card,7502.00
108,I338966,C907582,Clothing,5,1500.40,21-09-2022,Mall of Istanbul,Female,57.0,Cash,7502.00
129,I160221,C319231,Clothing,3,900.24,01-08-2022,Mall of Istanbul,Male,25.0,Cash,2700.72
138,I657069,C651699,Clothing,4,1200.32,16-12-2021,Mall of Istanbul,Male,40.0,Cash,4801.28


**Q15. Filter:**
- age is greater than 30 and price is above $500.
- category is "Shoes" and shopping_mall is "Dubai Mall".

In [54]:
all_data[( all_data['age'] > 30 ) & (all_data['price'] > 500)]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card,15004.25
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul,Female,32.0,Credit Card,1200.32
8,I293455,C159642,Clothing,3,900.24,04-11-2021,Metrocity,Male,69.0,Credit Card,2700.72
9,I326945,C283361,Clothing,2,600.16,22-08-2021,Kanyon,Female,60.0,Credit Card,1200.32
14,I336189,C555402,Clothing,2,600.16,03-06-2022,Kanyon,Female,67.0,Credit Card,1200.32
...,...,...,...,...,...,...,...,...,...,...,...
99435,I141871,C298790,Clothing,3,900.24,02-02-2022,Emaar Square Mall,Female,62.0,Cash,2700.72
99436,I579905,C265057,Clothing,5,1500.40,27-03-2022,Cevahir AVM,Female,57.0,Cash,7502.00
99442,I105959,C131810,Clothing,4,1200.32,01-02-2021,Istinye Park,Female,45.0,Debit Card,4801.28
99448,I332105,C231387,Shoes,4,2400.68,29-08-2021,Metropol AVM,Female,65.0,Credit Card,9602.72


In [55]:
all_data[(all_data['category'] == 'Shoes') & (all_data['shopping_mall'] == "Dubai Mall")]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending


**Q16. Retrieve transactions where:**

- The payment_method is "Cash".
- The quantity is more than 5 and the total spending exceeds $1000.

In [56]:
all_data[all_data['payment_method'] == 'Cash']

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash,300.08
4,I337046,C189076,Books,4,60.60,24-10-2021,Kanyon,Female,53.0,Cash,242.40
6,I121056,C151197,Cosmetics,1,40.66,13-03-2022,Istinye Park,Female,49.0,Cash,40.66
10,I306368,C240286,Food & Beverage,2,10.46,25-12-2022,Metrocity,Female,36.0,Cash,20.92
13,I179802,C312861,Clothing,2,600.16,17-11-2022,Cevahir AVM,Male,25.0,Cash,1200.32
...,...,...,...,...,...,...,...,...,...,...,...
99449,I134399,C953724,Clothing,1,300.08,01-01-2023,Kanyon,Male,65.0,Cash,300.08
99450,I170504,C226974,Books,1,15.15,28-02-2023,Zorlu Center,Female,28.0,Cash,15.15
99451,I675411,C513603,Toys,5,179.20,09-10-2021,Metropol AVM,Male,50.0,Cash,896.00
99453,I325143,C569580,Food & Beverage,2,10.46,22-09-2021,Forum Istanbul,Male,27.0,Cash,20.92


In [60]:
all_data[(all_data['quantity'] > 5) & (all_data['total_spending'] > 1000)]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending


**Q17. Group sales by category and calculate:**
- Total sales revenue for each category.
- Total number of items sold per category.

In [68]:
all_data.groupby('category')['total_spending'].sum()

category
Books              8.345529e+05
Clothing           1.139968e+08
Cosmetics          6.792863e+06
Food & Beverage    8.495351e+05
Shoes              6.655345e+07
Souvenir           6.357191e+05
Technology         5.786235e+07
Toys               3.980426e+06
Name: total_spending, dtype: float64

In [69]:
all_data.groupby('category')['quantity'].sum()

category
Books               14982
Clothing           103558
Cosmetics           45465
Food & Beverage     44277
Shoes               30217
Souvenir            14871
Technology          15021
Toys                30321
Name: quantity, dtype: int64

**Q18. Create a new column, age_group, categorizing customers as:
"Under 25", "25-40", and "Above 40".
Calculate the total spending for each group using the sum function.**

In [70]:
all_data['age_group']= 'age'

In [72]:
all_data.tail(1)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
99456,I232867,C273973,Souvenir,3,35.19,15-10-2022,Mall of Istanbul,Female,36.0,Credit Card,0.0,age


In [82]:
for i in range(0,99457):
    if all_data.loc[i,'age'] < 25:
        all_data.loc[i,'age_group'] = "Under 25"
    elif all_data.loc[i,'age'] > 24 and all_data.loc[i,'age'] < 41:
        all_data.loc[i,'age_group'] = "25-40"
    elif all_data.loc[i,'age'] > 40:
        all_data.loc[i,'age_group'] = "Above 40"
    else:
        all_data.loc[i,'age_group'] = np.nan

In [83]:
all_data

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card,7502.00,25-40
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul,Male,21.0,Debit Card,5401.53,Under 25
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash,300.08,Under 25
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card,15004.25,Above 40
4,I337046,C189076,Books,4,60.60,24-10-2021,Kanyon,Female,53.0,Cash,242.40,Above 40
...,...,...,...,...,...,...,...,...,...,...,...,...
99452,I219422,C441542,Souvenir,5,58.65,21-09-2022,Kanyon,Female,45.0,Credit Card,293.25,Above 40
99453,I325143,C569580,Food & Beverage,2,10.46,22-09-2021,Forum Istanbul,Male,27.0,Cash,20.92,25-40
99454,I824010,C103292,Food & Beverage,2,10.46,28-03-2021,Metrocity,Male,63.0,Debit Card,20.92,Above 40
99455,I702964,C800631,Technology,4,4200.00,16-03-2021,Istinye Park,Male,56.0,Cash,16800.00,Above 40


In [84]:
all_data.groupby("age_group")['total_spending'].sum()

age_group
25-40       7.786474e+07
Above 40    1.399327e+08
Under 25    3.343094e+07
Name: total_spending, dtype: float64

**Q19. Select rows where category is "Cosmetics" and shopping_mall is "Mall of Istanbul".
Find all customers who are over 30 years old and made purchases of "Clothing".**

In [86]:
all_data[(all_data['category'] == "Cosmetics") & (all_data['shopping_mall'] == "Mall of Istanbul")]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
26,I196704,C289625,Cosmetics,5,203.30,20-04-2022,Mall of Istanbul,Female,25.0,Credit Card,1016.50,25-40
46,I267193,C317818,Cosmetics,3,121.98,12-01-2023,Mall of Istanbul,Female,19.0,Credit Card,365.94,Under 25
178,I165202,C617189,Cosmetics,4,162.64,21-11-2022,Mall of Istanbul,Female,56.0,Cash,650.56,Above 40
185,I334246,C311034,Cosmetics,3,121.98,19-05-2022,Mall of Istanbul,Female,23.0,Cash,365.94,Under 25
223,I525798,C581947,Cosmetics,4,162.64,03-02-2023,Mall of Istanbul,Male,46.0,Cash,650.56,Above 40
...,...,...,...,...,...,...,...,...,...,...,...,...
99283,I133077,C876804,Cosmetics,5,203.30,23-12-2022,Mall of Istanbul,Female,46.0,Credit Card,1016.50,Above 40
99396,I628382,C826344,Cosmetics,1,40.66,19-10-2022,Mall of Istanbul,Male,56.0,Credit Card,40.66,Above 40
99428,I184085,C175610,Cosmetics,3,121.98,11-12-2021,Mall of Istanbul,Female,33.0,Credit Card,365.94,25-40
99431,I191779,C131280,Cosmetics,4,162.64,28-12-2021,Mall of Istanbul,Male,26.0,Credit Card,650.56,25-40


In [87]:
all_data[(all_data['age'] > 30) & ( all_data['category'] == "Clothing")]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul,Female,32.0,Credit Card,1200.32,25-40
8,I293455,C159642,Clothing,3,900.24,04-11-2021,Metrocity,Male,69.0,Credit Card,2700.72,Above 40
9,I326945,C283361,Clothing,2,600.16,22-08-2021,Kanyon,Female,60.0,Credit Card,1200.32,Above 40
14,I336189,C555402,Clothing,2,600.16,03-06-2022,Kanyon,Female,67.0,Credit Card,1200.32,Above 40
18,I993048,C218149,Clothing,2,600.16,26-07-2021,Metropol AVM,Female,46.0,Cash,1200.32,Above 40
...,...,...,...,...,...,...,...,...,...,...,...,...
99435,I141871,C298790,Clothing,3,900.24,02-02-2022,Emaar Square Mall,Female,62.0,Cash,2700.72,Above 40
99436,I579905,C265057,Clothing,5,1500.40,27-03-2022,Cevahir AVM,Female,57.0,Cash,7502.00,Above 40
99440,I887161,C137631,Clothing,1,300.08,22-06-2021,Kanyon,Female,58.0,Cash,300.08,Above 40
99442,I105959,C131810,Clothing,4,1200.32,01-02-2021,Istinye Park,Female,45.0,Debit Card,4801.28,Above 40


**Q20. How can you group merged data by both gender and category to calculate total and
average sales for each combination?**

In [90]:
all_data.groupby(['gender','category'])['total_spending'].sum()

gender  category       
Female  Books                489314.70
        Clothing           68251695.60
        Cosmetics           4066772.54
        Food & Beverage      505322.60
        Shoes              39425167.30
        Souvenir             382761.63
        Technology         34669950.00
        Toys                2416046.08
Male    Books                345238.20
        Clothing           45745095.44
        Cosmetics           2726090.36
        Food & Beverage      344212.45
        Shoes              27128284.17
        Souvenir             252957.45
        Technology         23192400.00
        Toys                1564380.16
Name: total_spending, dtype: float64

In [91]:
all_data.groupby(['gender','category'])['total_spending'].mean()

gender  category       
Female  Books                168.380833
        Clothing            3304.846775
        Cosmetics            448.376245
        Food & Beverage       57.396933
        Shoes               6607.200821
        Souvenir             126.868290
        Technology         11630.308621
        Toys                 397.049479
Male    Books                166.379855
        Clothing            3306.475999
        Cosmetics            452.312985
        Food & Beverage       57.637718
        Shoes               6670.342801
        Souvenir             127.627371
        Technology         11509.875931
        Toys                 390.899590
Name: total_spending, dtype: float64

**Q21. How can you create a crosstab to display the count for each combination of quantity and
category?**

In [94]:
pd.crosstab(all_data["quantity"],all_data["category"],margins=True)

category,Books,Clothing,Cosmetics,Food & Beverage,Shoes,Souvenir,Technology,Toys,All
quantity,,,,,,,,,
1,985,6841,2986,3002,1944,1024,985,2000,19767
2,1008,6950,2974,2909,2045,984,1000,1958,19828
3,1000,6907,3072,2992,2010,1077,1009,2082,20149
4,959,6849,3010,2884,2022,922,1001,2076,19723
5,1029,6940,3055,2989,2013,992,1001,1971,19990
All,4981,34487,15097,14776,10034,4999,4996,10087,99457


**Q22. How would you use a crosstab to display the total sales for each age group?**

In [101]:
pd.crosstab(all_data['age_group'],columns='Total Sales',values=all_data['total_spending'],aggfunc='sum')

col_0,Total Sales
age_group,
25-40,7.786474e+07
Above 40,1.399327e+08
Under 25,3.343094e+07


**Q23. How can you use a crosstab to show the total number of purchases for each category by
payment_method ?**

In [105]:
pd.crosstab(all_data['category'], all_data['payment_method'])

payment_method,Cash,Credit Card,Debit Card
category,,,
Books,2268,1696,1017
Clothing,15456,12025,7006
Cosmetics,6674,5336,3087
Food & Beverage,6587,5250,2939
Shoes,4474,3553,2007
Souvenir,2211,1770,1018
Technology,2235,1753,1008
Toys,4542,3548,1997


**Q24. How can you create a crosstab to analyze the number of purchases by gender for each
category?**

In [106]:
pd.crosstab(all_data['gender'],all_data['category'])

category,Books,Clothing,Cosmetics,Food & Beverage,Shoes,Souvenir,Technology,Toys
gender,,,,,,,,
Female,2906,20652,9070,8804,5967,3017,2981,6085
Male,2075,13835,6027,5972,4067,1982,2015,4002


**Q25. How would you create a crosstab to display the average quantity purchased for each
category by age_group?**

In [109]:
pd.crosstab(all_data['age_group'],all_data['category'],values=all_data['quantity'],aggfunc='mean')

category,Books,Clothing,Cosmetics,Food & Beverage,Shoes,Souvenir,Technology,Toys
age_group,,,,,,,,
25-40,3.034414,2.989489,3.030736,3.016009,3.029085,2.987890,2.988132,2.973988
Above 40,2.968323,3.019206,2.998812,2.986258,2.993918,2.955669,3.030226,3.017831
Under 25,3.106259,2.967094,3.020728,2.994149,3.042491,3.016418,2.953704,3.034164


**Q26. How can you use a crosstab to show the total spending (total_spending) by age_group and
payment_method?**

In [112]:
pd.crosstab(all_data['age_group'],all_data['payment_method'],values=all_data["total_spending"],aggfunc='sum')

payment_method,Cash,Credit Card,Debit Card
age_group,,,
25-40,35548808.07,27008420.64,15307511.81
Above 40,62291566.16,49118771.06,28522409.24
Under 25,14867543.65,11860635.02,6702761.56


**Q27. How can you filter data to show customers in the 25-40 age group who paid via Credit
Card?**

In [114]:
all_data[(all_data['age_group'] == "25-40") & (all_data['payment_method'] == "Credit Card")]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card,7502.00,25-40
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card,7502.00,25-40
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul,Female,32.0,Credit Card,1200.32,25-40
11,I139207,C191708,Books,1,15.15,28-10-2022,Emaar Square Mall,Female,29.0,Credit Card,15.15,25-40
26,I196704,C289625,Cosmetics,5,203.30,20-04-2022,Mall of Istanbul,Female,25.0,Credit Card,1016.50,25-40
...,...,...,...,...,...,...,...,...,...,...,...,...
99420,I278400,C281528,Cosmetics,5,203.30,12-05-2022,Istinye Park,Female,34.0,Credit Card,1016.50,25-40
99428,I184085,C175610,Cosmetics,3,121.98,11-12-2021,Mall of Istanbul,Female,33.0,Credit Card,365.94,25-40
99431,I191779,C131280,Cosmetics,4,162.64,28-12-2021,Mall of Istanbul,Male,26.0,Credit Card,650.56,25-40
99433,I403904,C142643,Toys,1,35.84,02-11-2022,Metrocity,Male,28.0,Credit Card,35.84,25-40


**Q28. How can you use the filter method to display transactions where the total spending
(total_spending) is greater than 1000?**

In [116]:
all_data.groupby('customer_id').filter(lambda x: x['total_spending'] > 1000)

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card,7502.00,25-40
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul,Male,21.0,Debit Card,5401.53,Under 25
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card,15004.25,Above 40
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card,7502.00,25-40
7,I293112,C176086,Clothing,2,600.16,13-01-2021,Mall of Istanbul,Female,32.0,Credit Card,1200.32,25-40
...,...,...,...,...,...,...,...,...,...,...,...,...
99435,I141871,C298790,Clothing,3,900.24,02-02-2022,Emaar Square Mall,Female,62.0,Cash,2700.72,Above 40
99436,I579905,C265057,Clothing,5,1500.40,27-03-2022,Cevahir AVM,Female,57.0,Cash,7502.00,Above 40
99442,I105959,C131810,Clothing,4,1200.32,01-02-2021,Istinye Park,Female,45.0,Debit Card,4801.28,Above 40
99448,I332105,C231387,Shoes,4,2400.68,29-08-2021,Metropol AVM,Female,65.0,Credit Card,9602.72,Above 40


**Q29. Display data for only the Credit Card and Cash payment methods?**

In [118]:
all_data[(all_data['payment_method'] == 'Cash') | (all_data['payment_method'] == 'Credit Card')]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card,7502.00,25-40
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash,300.08,Under 25
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card,15004.25,Above 40
4,I337046,C189076,Books,4,60.60,24-10-2021,Kanyon,Female,53.0,Cash,242.40,Above 40
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card,7502.00,25-40
...,...,...,...,...,...,...,...,...,...,...,...,...
99451,I675411,C513603,Toys,5,179.20,09-10-2021,Metropol AVM,Male,50.0,Cash,896.00,Above 40
99452,I219422,C441542,Souvenir,5,58.65,21-09-2022,Kanyon,Female,45.0,Credit Card,293.25,Above 40
99453,I325143,C569580,Food & Beverage,2,10.46,22-09-2021,Forum Istanbul,Male,27.0,Cash,20.92,25-40
99455,I702964,C800631,Technology,4,4200.00,16-03-2021,Istinye Park,Male,56.0,Cash,16800.00,Above 40


**30. Display data for only the Clothing and Shoes categories from the dataset?**

In [119]:
all_data[(all_data['category'] == 'Clothing') | (all_data['category'] == 'Shoes')]

,invoice_no,customer_id,category,quantity,price,invoice_date,shopping_mall,gender,age,payment_method,total_spending,age_group
0,I138884,C241288,Clothing,5,1500.40,05-08-2022,Kanyon,Female,28.0,Credit Card,7502.00,25-40
1,I317333,C111565,Shoes,3,1800.51,12-12-2021,Forum Istanbul,Male,21.0,Debit Card,5401.53,Under 25
2,I127801,C266599,Clothing,1,300.08,09-11-2021,Metrocity,Male,20.0,Cash,300.08,Under 25
3,I173702,C988172,Shoes,5,3000.85,16-05-2021,Metropol AVM,Female,66.0,Credit Card,15004.25,Above 40
5,I227836,C657758,Clothing,5,1500.40,24-05-2022,Forum Istanbul,Female,28.0,Credit Card,7502.00,25-40
...,...,...,...,...,...,...,...,...,...,...,...,...
99436,I579905,C265057,Clothing,5,1500.40,27-03-2022,Cevahir AVM,Female,57.0,Cash,7502.00,Above 40
99440,I887161,C137631,Clothing,1,300.08,22-06-2021,Kanyon,Female,58.0,Cash,300.08,Above 40
99442,I105959,C131810,Clothing,4,1200.32,01-02-2021,Istinye Park,Female,45.0,Debit Card,4801.28,Above 40
99448,I332105,C231387,Shoes,4,2400.68,29-08-2021,Metropol AVM,Female,65.0,Credit Card,9602.72,Above 40
